mlir/include/mlir/Bytecode/BytecodeWriter.h

```diff
@@ -97,6 +98,14 @@ class BytecodeWriterConfig {
   /// Get the set desired bytecode version to emit.
   int64_t getDesiredBytecodeVersion() const;
 
+  /// A map containing the dialect versions to emit.
+  llvm::StringMap<std::unique_ptr<DialectVersion>> &
+  getDialectVersionMap() const;
```

Comment:
Can this be a `DenseMap<Dialect *, <std::unique_ptr<DialectVersion>> &`?
That is: do we have cases for getting this information from a place without a MLIRContext or without the ability to get a `Dialect*`?

Reply:
I mostly left the string map to be consistent to the reader API (the string there makes sense since you need to trigger loading the dialect in the context). The only case I can think of that would require having a string map involves callbacks - for example, what if you want to apply a version to a dialect that does not exist in memory but you still want to back deploy to? 

Although possible, I don't think it will be a common case though - mostly you will query the version within BytecodeDialectInterface, which means you probably have the dialect loaded in the context already. I am not sure if my previous argument warrant keeping it as a string map.

Reply:
And this is to allow reading multiple versions? E.g., why not hide the map as an internal detail and require dialect as arg, and just have it return a DialectVersion ?

Reply:
The map is referenced within the BytecodeDialectWriter, so this is mostly convenience - the alternative is to pass the writer configs to the BytecodeDialectWriter itself. Reading multiple versions is really needed for the callback-mechanism (where you could use version X to conditionally emit encodings for dialect Y). For the other cases you should not need to query registered versions other than `self`.